llvm/lib/Target/AIE/AIE2AsmPrinter.cpp

```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
   return LowerAIEMachineOperandToMCOperand(MO, MCOp, *this);
 }
 
+void AIE2AsmPrinter::emitInstruction(const MachineInstr *MI) {
+  // PseudoLoopEnd is just here to carry some extra information. No need to
+  // emit anything
+  if (MI->getOpcode() == AIE2::PseudoLoopEnd)
+    return;
+
+  // Check whether there's a PseudoLoopEnd following. If so, emit the
+  // last-bundle-label that it carries to designate this MI as the last bundle.
+  const MachineInstr *LastBundle = nullptr;
+  const MachineInstr *LoopEnd = nullptr;
+  const MachineInstr *Prev = nullptr;
+  for (auto &MII : *MI->getParent()) {
+    if (MII.getOpcode() == AIE2::PseudoLoopEnd) {
+      LoopEnd = &MII;
+      LastBundle = Prev;
+      break;
+    }
+    Prev = &MII;
+  }
```

Comment:
I think that analysis could be done a single time before emitting any instruction is called. If we can't use pre/post emission symbols, maybe move that analysis into e.g. `AIEBaseAsmPrinter::emitFunctionBodyStart`?